src/features/broker-protection/actions/extract.js

```diff
@@ -0,0 +1,300 @@
+ *       "city": "Orlando",
+ *       "state": "FL"
+ *     }
+ *   ],
+ *   "profileUrl": "https://example.com/1234"
+ * }
+ *
+ * @param {HTMLElement} profileElement
+ * @param {Record<string, ExtractProfileProperty>} extractData
+ * @return {Record<string, any>}
+ */
+function createProfile (profileElement, extractData) {
+    const output = {}
+    for (const [key, value] of Object.entries(extractData)) {
+        if (!value?.selector) {
+            output[key] = null
+        } else {
+            const evaluatedValue = value?.findElements
+                ? findFromElements(profileElement, key, value)
+                : findFromElement(profileElement, key, value)
+
+            const extractedValue = extractValue(key, value, evaluatedValue)
+            // counts an empty string as null since that is essentially a blank extracted value
+            output[key] = extractedValue || evaluatedValue || null
+        }
+    }
+    return output
+}
+
+/**
+ * @param {HTMLElement} profileElement
+ * @param {string} key
+ * @param {ExtractProfileProperty} extractField
+ */
+function findFromElements (profileElement, key, extractField) {
+    const elements = getElements(profileElement, extractField.selector) || null
+    const elementValues = []
+    if (elements) {
+        for (const element of elements) {
+            let elementValue = rules[key]?.(element) ?? element?.innerText ?? null
+            if (extractField?.afterText) {
+                elementValue = elementValue?.split(extractField.afterText)[1]?.trim() || elementValue
+            }
+            if (extractField?.beforeText) {
+                elementValue = elementValue?.split(extractField.beforeText)[0].trim() || elementValue
+            }
+            elementValues.push(elementValue)
+        }
+    }
+    return elementValues
+}
+
+/**
+ * @param {HTMLElement} profileElement
+ * @param {string} dataKey - such as 'name', 'age' etc
+ * @param {ExtractProfileProperty} extractField
+ * @return {string}
+ */
+function findFromElement (profileElement, dataKey, extractField) {
+    const element = getElement(profileElement, extractField.selector) ||
+        getElementMatches(profileElement, extractField.selector)
+
+    // todo: should we use textContent here?
+    let elementValue = rules[dataKey]?.(element) ?? element?.innerText ?? null
+
+    if (extractField?.afterText) {
+        elementValue = elementValue?.split(extractField.afterText)[1]?.trim() || elementValue
+    }
+    // there is a case where we may want to get the text "after" and "before" certain text
+    if (extractField?.beforeText) {
+        elementValue = elementValue?.split(extractField.beforeText)[0].trim() || elementValue
+    }
+    return elementValue
+}
+
+/**
+ * Try to filter partial data based on the user's actual profile data
+ * @param {Record<string, any>} userData
+ * @param {Record<string, any>} scrapedData
+ * @return {boolean}
+ */
+function scrapedDataMatchesUserData (userData, scrapedData) {
```

Comment:
notice: I just glanced over this and underlying functions as they are largely heuristics and data massage, so I can't really judge the correctness